File: forms/anesthesiology-assessment/full-stack-with-loco-tera-htmx-alpine/tests/engine/composite_grader_test.rs
use anesthesiology_assessment_tera_crate::engine::asa_rules::asa_risk_from_class;
use anesthesiology_assessment_tera_crate::engine::composite_grader::grade;
use anesthesiology_assessment_tera_crate::engine::rcri_rules::{
    rcri_mace_percent, rcri_risk_from_score,
};
use anesthesiology_assessment_tera_crate::engine::stopbang_rules::stopbang_risk_from_score;
use anesthesiology_assessment_tera_crate::engine::types::*;

#[test]
fn empty_assessment_is_low_risk() {
    let data = AssessmentData::default();
    let r = grade(&data);
    assert_eq!(r.overall_risk, "low");
    assert_eq!(r.asa.risk_level, "low");
    assert_eq!(r.airway.risk_level, "low");
    assert_eq!(r.rcri.score, 0);
    assert_eq!(r.stopbang.score, 0);
}

#[test]
fn asa_iv_promotes_overall_to_high() {
    let mut data = AssessmentData::default();
    data.investigations_and_plan.asa_class = "iv".to_string();
    let r = grade(&data);
    assert_eq!(r.asa.risk_level, "high");
    assert_eq!(r.overall_risk, "high");
}

#[test]
fn asa_v_promotes_overall_to_critical() {
    let mut data = AssessmentData::default();
    data.investigations_and_plan.asa_class = "v".to_string();
    let r = grade(&data);
    assert_eq!(r.asa.risk_level, "critical");
    assert_eq!(r.overall_risk, "critical");
}

#[test]
fn asa_risk_mapping() {
    assert_eq!(asa_risk_from_class("i"), "low");
    assert_eq!(asa_risk_from_class("ii"), "low");
    assert_eq!(asa_risk_from_class("iii"), "medium");
    assert_eq!(asa_risk_from_class("iv"), "high");
    assert_eq!(asa_risk_from_class("v"), "critical");
    assert_eq!(asa_risk_from_class("vi"), "critical");
    assert_eq!(asa_risk_from_class(""), "low");
}

#[test]
fn mallampati_iv_is_high_airway_risk() {
    let mut data = AssessmentData::default();
    data.physical_exam.mallampati_class = "iv".to_string();
    let r = grade(&data);
    assert_eq!(r.airway.risk_level, "high");
    assert!(r.airway.fired_rules.iter().any(|f| f.id == "MALL-002"));
}

#[test]
fn difficult_intubation_history_is_high_airway_risk() {
    let mut data = AssessmentData::default();
    data.previous_anaesthesia.difficult_intubation = true;
    let r = grade(&data);
    assert_eq!(r.airway.risk_level, "high");
    assert!(r.airway.fired_rules.iter().any(|f| f.id == "MALL-008"));
}

#[test]
fn two_medium_airway_factors_promote_to_high() {
    let mut data = AssessmentData::default();
    data.physical_exam.mallampati_class = "iii".to_string();
    data.physical_exam.mouth_opening = Some(2.5);
    let r = grade(&data);
    assert_eq!(r.airway.risk_level, "high");
    assert_eq!(r.airway.medium_factors, 2);
}

#[test]
fn rcri_three_criteria_critical() {
    let mut data = AssessmentData::default();
    data.investigations_and_plan.rcri_ischaemic_heart_disease = "yes".to_string();
    data.investigations_and_plan.rcri_congestive_heart_failure = "yes".to_string();
    data.investigations_and_plan.rcri_cerebrovascular_disease = "yes".to_string();
    let r = grade(&data);
    assert_eq!(r.rcri.score, 3);
    assert_eq!(r.rcri.risk_level, "critical");
    assert!((r.rcri.mace_percent - 5.4).abs() < 0.0001);
}

#[test]
fn rcri_high_risk_surgery_counts() {
    let mut data = AssessmentData::default();
    data.planned_surgery.surgery_grade = "major".to_string();
    let r = grade(&data);
    assert_eq!(r.rcri.score, 1);
    assert_eq!(r.rcri.risk_level, "medium");
}

#[test]
fn rcri_risk_mapping() {
    assert_eq!(rcri_risk_from_score(0), "low");
    assert_eq!(rcri_risk_from_score(1), "medium");
    assert_eq!(rcri_risk_from_score(2), "high");
    assert_eq!(rcri_risk_from_score(3), "critical");
    assert_eq!(rcri_risk_from_score(6), "critical");
}

#[test]
fn rcri_mace_percent_mapping() {
    assert!((rcri_mace_percent(0) - 0.4).abs() < 0.0001);
    assert!((rcri_mace_percent(1) - 1.0).abs() < 0.0001);
    assert!((rcri_mace_percent(2) - 2.4).abs() < 0.0001);
    assert!((rcri_mace_percent(3) - 5.4).abs() < 0.0001);
}

#[test]
fn stopbang_five_items_is_high() {
    let mut data = AssessmentData::default();
    data.social_history.snores_loudly = "yes".to_string();
    data.social_history.tired_during_day = "yes".to_string();
    data.social_history.observed_apnea = "yes".to_string();
    data.medical_history.hypertension = "yes".to_string();
    data.vital_signs.bmi = Some(36.0);
    let r = grade(&data);
    assert_eq!(r.stopbang.score, 5);
    assert_eq!(r.stopbang.risk_level, "high");
}

#[test]
fn stopbang_mapping() {
    assert_eq!(stopbang_risk_from_score(0), "low");
    assert_eq!(stopbang_risk_from_score(2), "low");
    assert_eq!(stopbang_risk_from_score(3), "medium");
    assert_eq!(stopbang_risk_from_score(4), "medium");
    assert_eq!(stopbang_risk_from_score(5), "high");
    assert_eq!(stopbang_risk_from_score(8), "high");
}

#[test]
fn worst_risk_wins_for_overall() {
    let mut data = AssessmentData::default();
    // ASA III = medium
    data.investigations_and_plan.asa_class = "iii".to_string();
    // RCRI 3 = critical
    data.investigations_and_plan.rcri_ischaemic_heart_disease = "yes".to_string();
    data.investigations_and_plan.rcri_congestive_heart_failure = "yes".to_string();
    data.investigations_and_plan.rcri_cerebrovascular_disease = "yes".to_string();
    let r = grade(&data);
    assert_eq!(r.overall_risk, "critical");
}

File: forms/anesthesiology-assessment/front-end-with-svelte/src/lib/engine/composite-grader.test.ts
import { describe, it, expect } from 'vitest';
import { gradeAssessment } from './composite-grader';
import { detectAdditionalFlags } from './flagged-issues';
import { evaluateAsa, asaRiskFromClass } from './asa-rules';
import { evaluateAirway } from './mallampati-rules';
import { evaluateRcri, rcriRiskFromScore } from './rcri-rules';
import { evaluateStopbang, stopbangRiskFromScore } from './stopbang-rules';
import { createDefaultAssessment } from '#lib/stores/assessment.svelte.js';
import type { AssessmentData } from './types';

function healthyPatient(): AssessmentData {
	const d = createDefaultAssessment();
	d.demographics.firstName = 'Test';
	d.demographics.lastName = 'Patient';
	d.demographics.dateOfBirth = '1990-01-01';
	d.demographics.sex = 'female';
	d.investigationsAndPlan.asaClass = 'i';
	return d;
}

describe('ASA sub-grader', () => {
	it('maps ASA classes to risk bands', () => {
		expect(asaRiskFromClass('i')).toBe('low');
		expect(asaRiskFromClass('ii')).toBe('low');
		expect(asaRiskFromClass('iii')).toBe('medium');
		expect(asaRiskFromClass('iv')).toBe('high');
		expect(asaRiskFromClass('v')).toBe('critical');
		expect(asaRiskFromClass('vi')).toBe('critical');
		expect(asaRiskFromClass('')).toBe('low');
	});

	it('fires a rule and flags emergency cases', () => {
		const d = healthyPatient();
		d.investigationsAndPlan.asaClass = 'iv';
		d.investigationsAndPlan.emergencyCase = 'yes';
		const r = evaluateAsa(d);
		expect(r.riskLevel).toBe('high');
		expect(r.emergency).toBe(true);
		expect(r.firedRules).toHaveLength(1);
		expect(r.firedRules[0].description).toContain('Emergency');
	});
});

describe('Mallampati / airway sub-grader', () => {
	it('returns low risk for an unremarkable airway', () => {
		const d = healthyPatient();
		d.physicalExam.mallampatiClass = 'ii';
		expect(evaluateAirway(d).riskLevel).toBe('low');
	});

	it('returns high risk for Mallampati IV', () => {
		const d = healthyPatient();
		d.physicalExam.mallampatiClass = 'iv';
		expect(evaluateAirway(d).riskLevel).toBe('high');
	});

	it('escalates to high risk with two medium factors', () => {
		const d = healthyPatient();
		d.physicalExam.mallampatiClass = 'iii';
		d.physicalExam.mouthOpening = 2.5;
		const r = evaluateAirway(d);
		expect(r.mediumFactors).toBe(2);
		expect(r.riskLevel).toBe('high');
	});

	it('returns high risk for a previous difficult intubation', () => {
		const d = healthyPatient();
		d.previousAnaesthesia.difficultIntubation = true;
		expect(evaluateAirway(d).riskLevel).toBe('high');
	});
});

describe('RCRI sub-grader', () => {
	it('maps scores to risk bands', () => {
		expect(rcriRiskFromScore(0)).toBe('low');
		expect(rcriRiskFromScore(1)).toBe('medium');
		expect(rcriRiskFromScore(2)).toBe('high');
		expect(rcriRiskFromScore(3)).toBe('critical');
	});

	it('counts criteria and computes MACE percent', () => {
		const d = healthyPatient();
		d.plannedSurgery.surgeryGrade = 'major';
		d.investigationsAndPlan.rcriIschaemicHeartDisease = 'yes';
		d.investigationsAndPlan.rcriCongestiveHeartFailure = 'yes';
		const r = evaluateRcri(d);
		expect(r.score).toBe(3);
		expect(r.riskLevel).toBe('critical');
		expect(r.macePercent).toBe(5.4);
	});
});

describe('STOP-BANG sub-grader', () => {
	it('maps scores to risk bands', () => {
		expect(stopbangRiskFromScore(2)).toBe('low');
		expect(stopbangRiskFromScore(3)).toBe('medium');
		expect(stopbangRiskFromScore(5)).toBe('high');
	});

	it('counts the eight items', () => {
		const d = healthyPatient();
		d.demographics.sex = 'male';
		d.demographics.dateOfBirth = '1950-01-01';
		d.medicalHistory.hypertension = 'yes';
		d.socialHistory.snoresLoudly = 'yes';
		d.socialHistory.tiredDuringDay = 'yes';
		d.vitalSigns.bmi = 36;
		d.vitalSigns.neckCircumference = 43;
		const r = evaluateStopbang(d);
		expect(r.score).toBe(7);
		expect(r.riskLevel).toBe('high');
	});
});

describe('Composite grader', () => {
	it('returns low risk for a healthy ASA I patient', () => {
		const d = healthyPatient();
		const r = gradeAssessment(d);
		expect(r.overallRisk).toBe('low');
		expect(r.additionalFlags).toHaveLength(0);
	});

	it('promotes the worst sub-risk to the overall risk', () => {
		const d = healthyPatient();
		d.investigationsAndPlan.asaClass = 'ii'; // low
		d.physicalExam.mallampatiClass = 'iv'; // high
		const r = gradeAssessment(d);
		expect(r.overallRisk).toBe('high');
	});

	it('returns critical risk for ASA V', () => {
		const d = healthyPatient();
		d.investigationsAndPlan.asaClass = 'v';
		expect(gradeAssessment(d).overallRisk).toBe('critical');
	});

	it('collects fired rules from every sub-grader', () => {
		const d = healthyPatient();
		d.investigationsAndPlan.asaClass = 'iii';
		d.physicalExam.mallampatiClass = 'iii';
		d.plannedSurgery.surgeryGrade = 'major';
		d.socialHistory.snoresLoudly = 'yes';
		const r = gradeAssessment(d);
		const ids = r.firedRules.map((f) => f.id);
		expect(ids).toContain('ASA-001');
		expect(ids).toContain('MALL-001');
		expect(ids).toContain('RCRI-001');
		expect(ids).toContain('SB-S');
	});

	it('produces unique fired-rule ids', () => {
		const d = healthyPatient();
		d.investigationsAndPlan.asaClass = 'iii';
		d.physicalExam.mallampatiClass = 'iii';
		d.physicalExam.jawProtrusion = 'limited';
		d.plannedSurgery.surgeryGrade = 'major';
		d.socialHistory.snoresLoudly = 'yes';
		d.socialHistory.tiredDuringDay = 'yes';
		const r = gradeAssessment(d);
		const ids = r.firedRules.map((f) => f.id);
		expect(new Set(ids).size).toBe(ids.length);
	});
});

describe('Flagged issues', () => {
	it('returns no flags for a healthy patient', () => {
		expect(detectAdditionalFlags(healthyPatient())).toHaveLength(0);
	});

	it('flags latex allergy and anaphylaxis history', () => {
		const d = healthyPatient();
		d.allergies.latexAllergy = 'yes';
		d.allergies.list = [{ allergen: 'Penicillin', type: 'drug', reaction: 'Throat swelling', severity: 'anaphylaxis' }];
		const flags = detectAdditionalFlags(d);
		expect(flags.some((f) => f.id === 'FLAG-LATEX')).toBe(true);
		expect(flags.some((f) => f.id.startsWith('FLAG-ANAPH'))).toBe(true);
	});

	it('flags malignant hyperthermia risk', () => {
		const d = healthyPatient();
		d.previousAnaesthesia.malignantHyperthermia = 'yes';
		expect(detectAdditionalFlags(d).some((f) => f.id === 'FLAG-MH')).toBe(true);
	});

	it('flags anticoagulant use', () => {
		const d = healthyPatient();
		d.medications.onAnticoagulants = 'yes';
		expect(detectAdditionalFlags(d).some((f) => f.id === 'FLAG-ANTICOAG')).toBe(true);
	});

	it('sorts flags urgent first', () => {
		const d = healthyPatient();
		d.socialHistory.smoking = 'current'; // low
		d.allergies.list = [{ allergen: 'Latex', type: 'latex', reaction: 'Shock', severity: 'anaphylaxis' }]; // urgent
		const flags = detectAdditionalFlags(d);
		const order = { urgent: 0, high: 1, medium: 2, low: 3 };
		const priorities = flags.map((f) => order[f.priority]);
		expect(priorities).toEqual([...priorities].sort((a, b) => a - b));
	});
});
